src/libraries/System.Private.CoreLib/src/System/Text/Ascii.CaseConversion.cs

```diff
@@ -0,0 +1,605 @@
+            => ChangeCase<ushort, byte, ToLowerConversion>(MemoryMarshal.Cast<char, ushort>(source), destination, out bytesWritten);
+
+        /// <summary>
+        /// Performs in-place uppercase conversion.
+        /// </summary>
+        /// <param name="value">The ASCII text buffer.</param>
+        /// <param name="bytesWritten">The number of processed bytes.</param>
+        /// <returns>An <see cref="OperationStatus"/> describing the result of the operation.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static OperationStatus ToLowerInPlace(Span<byte> value, out int bytesWritten)
+            => ChangeCase<byte, ToLowerConversion>(value, out bytesWritten);
+
+        /// <summary>
+        /// Performs in-place uppercase conversion.
+        /// </summary>
+        /// <param name="value">The ASCII text buffer.</param>
+        /// <param name="charsWritten">The number of processed characters.</param>
+        /// <returns>An <see cref="OperationStatus"/> describing the result of the operation.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static OperationStatus ToLowerInPlace(Span<char> value, out int charsWritten)
+            => ChangeCase<ushort, ToLowerConversion>(MemoryMarshal.Cast<char, ushort>(value), out charsWritten);
+
+        /// <summary>
+        /// Performs in-place lowercase conversion.
+        /// </summary>
+        /// <param name="value">The ASCII text buffer.</param>
+        /// <param name="bytesWritten">The number of processed bytes.</param>
+        /// <returns>An <see cref="OperationStatus"/> describing the result of the operation.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static OperationStatus ToUpperInPlace(Span<byte> value, out int bytesWritten)
+            => ChangeCase<byte, ToUpperConversion>(value, out bytesWritten);
+
+        /// <summary>
+        /// Performs in-place lowercase conversion.
+        /// </summary>
+        /// <param name="value">The ASCII text buffer.</param>
+        /// <param name="charsWritten">The number of processed characters.</param>
+        /// <returns>An <see cref="OperationStatus"/> describing the result of the operation.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static OperationStatus ToUpperInPlace(Span<char> value, out int charsWritten)
+            => ChangeCase<ushort, ToUpperConversion>(MemoryMarshal.Cast<char, ushort>(value), out charsWritten);
+
+        private static unsafe OperationStatus ChangeCase<TFrom, TTo, TCasing>(ReadOnlySpan<TFrom> source, Span<TTo> destination, out int destinationElementsWritten)
+            where TFrom : unmanaged, IBinaryInteger<TFrom>
+            where TTo : unmanaged, IBinaryInteger<TTo>
+            where TCasing : struct
+        {
+            if (MemoryMarshal.AsBytes(source).Overlaps(MemoryMarshal.AsBytes(destination)))
```

Comment:
`TFROM` and `TTO` can be different (char and byte for example) and in theory someone could do sth like this:

https://github.com/dotnet/runtime/blob/c0f38d1689e4adedd661385ba8fe7c931fad3ba6/src/libraries/System.Text.Encoding/tests/Ascii/CaseConversionTests.cs#L28-L29

Since everything can be casted to bytes I decided to use it and cover all possible scenarios, but to be honest I did it mostly to cover all possible unit testing scenarios rather than expecting someone to actually do it.